Feat/code review assistant

immortal71 opened the pull request:
It includes : ~
--->Added a Code Review Assistant with AST-based static analysis and a PEP8 checker.
--->Files added: assistant + builder + CLI + Streamlit page + tools + schema + tests.
--->Acceptance: Pydantic schema, at least two tools, CLI, tests, builder pattern, optional Streamlit.
--->Tested: CLI and tool tests instructions and example code included in README.

immortal71 committed:
feat: added Code Review Assistant

immortal71 commented:
@nimonkaranurag  does this pr meet the requirement ?, does this need any modification , I it does just mention me I will do it ... I will wait for the review :)

nimonkaranurag commented on assistants/code_review_assistant/code_review_assistant_builder.py:
```diff
@@ -0,0 +1,64 @@
+        if not self.tools:
+            self.tools = [analyze_code, check_pep8]
+
+    def render_prompt(self) -> str:
```
I am not sure if I understand completely but what is the use of the llm here?

You are essentially implementing a rule-based traditional assistant, where are you using the llm for decision making?

immortal71 replied:
@nimonkaranurag  You’re right that in the current implementation most of the behavior is still rule-based: the builder defaults to a fixed set of tools (AST + PEP8), and the assistant would mostly work even without the LLM in the loop
I’m planning to:

--Refactor the assistant so the LLM acts as the central agent: it reads the user’s prompt, decides which tools to run (AST, PEP8, and additional tools I’ll add), and in what order.

--Use the LLM to synthesize the tool outputs into a higher‑level, opinionated code review with prioritized issues and concrete suggestions, rather than just dumping raw checker results.

--Improve the interface to be primarily natural‑language driven (and enhance the Streamlit UI) so it behaves more like a Copilot‑style code review assistant.
so -- can i proced this way or should i make some changes on my method, if u can just state what to do next i will work and update Asap!!! 🐧

nimonkaranurag reviewed (changes requested):
Hey @immortal71 sorry if the issue description wasn't clear enough/mis-leading.


Here are some concerns I would like to share:
- it's good see the distinction you have understood between an "agent" and an "LLM-agent", these are both valid "agents"(or, "assistants"). You don't need an LLM backend for decision making, rule-based agents also exist and this is a good use-case for it.
- I would try to see what features co-pilot brings to the table, as a starting point this is good but there is a lot more potential for enhancement.

Here are some TODOs I think will enhance your learning:
- [ ] the interface for interacting with the assistant shouldn't be a CLI tool but a natural language interface: "hey, can you make some recommendations for better engineering practices in this code?"
- [ ] integrate the LLM for decision making by:
  - [ ] expanding on current tools(add more tools that will be helpful to an agent in reviewing the code, let it pick the one(s) it needs at its discretion).
- [ ] add a `streamlit` front-end for your assistant, and add some images so I can see your integration! Will an "upload file" button work here?
- [ ] should you provide the assistant with some documentation or a web search tool for reading docs dynamically?
- [ ] think how you can make this assistant the one-stop solution for enhancing code.

Looking forward to hearing your thoughts!

immortal71 commented:
@nimonkaranurag   Thank you for the detailed feedback and suggestions! I’m excited to explore ways to improve the assistant-especially regarding the natural language interface and integrating LLMs for more dynamic decision-making .I’m aiming to deepen my understanding of open source contributions and would really appreciate any further advice you might have.
   If you have some time, I’d love to learn more from you about best practices or get your perspective on how I can improve as a contributor. I’m thinking about applying for GSoC next year but still not sure, so your mentorship or occasional guidance as I progress would mean a lot to me. And If you have any advice as a senior on org I would love to work on it too.
          I’m looking forward to collaborating more and growing along the way!

immortal71 committed:
refactor: transform code review assistant into LLM-driven agent

-- Added LLM-powered tool selection based on natural language queries
-- Implemented 4 new analysis tools: complexity, security, best practices, documentation search
-- Created Streamlit UI for interactive code reviews
-- Updated system prompt to guide intelligent agent behavior
-- Expanded test coverage for all new tools
- Document LLM-driven approach and usage examples

immortal71 commented:
@nimonkaranurag  can you check it, sorry i got little busy for past few days, and there is other issue that is under review too so can u check that too ?, u told me u will review later !!
